src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -511,6 +630,81 @@ public void SendRecvPollSync_TcpListener_Socket(IPAddress listenAt, bool pollBef
                 }
             }
         }
+
+        [Theory]
+        [InlineData(false, 1)]
+        [InlineData(true, 1)]
+        public async Task SendRecv_BlockingNonBlocking_LingerTimeout_Success(bool blocking, int lingerTimeout)
+        {
+            if (!SupportsNonBlocking) return;
+
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                client.Blocking = blocking;
+                listener.Blocking = blocking;
+
+                client.LingerState = new LingerOption(true, lingerTimeout);
+                listener.LingerState = new LingerOption(true, lingerTimeout);
+
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(1);
+
+                Task<Socket> acceptTask = AcceptAsync(listener);
+                await Task.WhenAll(
+                    acceptTask,
+                    ConnectAsync(client, new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port)));
+
+                using (Socket server = await acceptTask)
+                {
+                    server.Blocking = blocking;
+                    server.LingerState = new LingerOption(true, lingerTimeout);
+
+                    Task<int> receive = ReceiveAsync(client, new ArraySegment<byte>(new byte[1]));
+                    Assert.Equal(1, await SendAsync(server, new ArraySegment<byte>(new byte[1])));
+                    Assert.Equal(1, await receive);
+                }
+            }
+        }
+
+        [ActiveIssue(16716, TestPlatforms.OSX)] // SendBufferSize = 0 throws
+        [Fact]
+        public async Task SendRecv_NoBuffering_Success()
+        {
+            if (!SupportsNonBlocking) return;
+
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(1);
+
+                Task<Socket> acceptTask = AcceptAsync(listener);
+                await Task.WhenAll(
+                    acceptTask,
+                    ConnectAsync(client, new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port)));
+
+                client.SendBufferSize = 0;
```

Comment:
Should we have a similar test for ReceiveBufferSize?  Not sure it's legal to set that to 0, is it?